Improve vesting discovery and error messages

sebastinez opened the pull request:
When loading a ETH address as vesting contract we show a cryptic error
message, which non tech-savy end users don't understand, 8c847b0
changes it to something more human readable.

We allow users to enter and load a contract that probably isn't a
vesting contract, we show this now by showing an error message.

The `getInfo` function in the `vesting` folder is very prone to possible
errors, and benefits from some parallelization between all the method
calls, and some error handling.

We allow hereby to enter a contract address in search and navigate to the
contract profile page where we provide a button to display the detailed vesting
information if available.

On the same profile page, the user has the possibility to withdraw their withdrawable balance
also the balance is being increased on every new block.

sebastinez commented:
Here a video of some of the improvements:

Show a error message when trying to load a contract that probably isn't a vesting contract
<img width="842" alt="Bildschirm­foto 2022-12-21 um 18 25 23" src="https://user-images.githubusercontent.com/7912302/208966762-81749568-aba7-45f6-a911-eb3856bd7915.png">

Show a link to the vesting information on a contract profile page if available
We preload the vestingInfo so the loading time is instant.
<img width="953" alt="Bildschirm­foto 2022-12-21 um 18 27 43" src="https://user-images.githubusercontent.com/7912302/208967214-a3fcba6d-eb48-428f-89cf-435d4d9f856f.png">

If we try to lookup a vesting contract that doesn't exist with `/vesting/<contract address>` we show a `NotFound` component with some text for the user
<img width="947" alt="Bildschirm­foto 2022-12-21 um 18 29 59" src="https://user-images.githubusercontent.com/7912302/208967548-c482d3a7-430a-4236-98ee-b28c0d2f4a29.png">

rudolfs commented on public/typography.css:
```diff
@@ -137,6 +137,10 @@ p {
   color: var(--color-foreground-5);
   font-style: italic;
 }
+.txt-error {
```
We're just using this in a single place, maybe we can just inline it there?

sebastinez replied:
Yeah I wanted to keep the line of doing it, and eventually with more error handling, we could start using something like this. But for now we can do it inline

sebastinez replied:
634126e

rudolfs commented on src/App.svelte:
```diff
@@ -102,7 +102,7 @@
           activeRoute={$activeRouteStore} />
       {:else if $activeRouteStore.resource === "vesting"}
-        <Vesting {wallet} session={$session} activeRoute={$activeRouteStore} />
+        <Vesting {wallet} activeRoute={$activeRouteStore} />
```
Do we even want to keep the special vesting route now that we can just use the search bar?

sebastinez replied:
Yeah I wanted to remove it initially but some core contributors received mails with the /vesting route, to enter/load their contract.
So didn't want to break that.
Maybe at some point in the future?

rudolfs replied:
Or we could just show instructions on that page that say: please paste your vesting contract into the search bar at the top?

sebastinez replied:
Yeah that doesn't sound like good UX.
IMO either:
- we have a page like currently where we only show vesting information under a specific route.
- or we add it to a profile page like this PR wants to do it, and:
  -  show it in some nice condensed graphical easy to understand way with graphs maybe?
  -  or we hide it with a dropdown to avoid having lots of lines of information.

rudolfs commented:
Why do we have the vesting toggle button? Could we just show in the badge that it's a vesting contract?

<img width="899" alt="Screenshot 2022-12-22 at 18 02 39" src="https://user-images.githubusercontent.com/158411/209187885-ccf996a7-732f-4c2e-9f29-e7c0bc580113.png">

rudolfs commented:
Gonna tag @brandonhaslegs here, maybe he has some ideas on the design. The interaction feels a bit weird to me, wonder if there's a better way? Also the link to Etherscan could be nicer.

https://user-images.githubusercontent.com/158411/209188684-db0bebf3-c91d-4249-8e0f-be34bbf1d52f.mov



<img width="936" alt="Screenshot 2022-12-22 at 18 04 28" src="https://user-images.githubusercontent.com/158411/209188283-9fb8fd1c-849a-4560-977c-cae88fa0250f.png">

sebastinez commented:
> Why do we have the vesting toggle button? Could we just show in the badge that it's a vesting contract?

Yeah this was the initial way I thought of doing it, but I didn't liked the amount of lines, showing up, so the idea of the vesting toggle button is to hide all that information behind a toggle.

@brandonhaslegs do you have any idea here maybe on how to make this nicer, I don't think we should show all the vesting information from the get go, since it uses lot of space.

rudolfs commented on src/base/vesting/vesting.ts:
```diff
@@ -37,50 +46,88 @@ export async function withdrawVested(
-  const tx = await contract.connect(signer).withdrawVested();
+    state.set({ type: "withdrawing", tx });
+    await tx.wait();
+  } catch (e: any) {
```
We shouldn't cast errors to `any`, by default it's `unknown`. Further down we should check that the error actually has a `.reason` attribute.

sebastinez replied:
28746cd should improve it

rudolfs commented on src/base/vesting/vesting.ts:
```diff
@@ -37,50 +46,88 @@ export async function withdrawVested(
+    state.set({ type: "withdrawing", tx });
+    await tx.wait();
+  } catch (e: any) {
+    console.log(e);
```
`console.warn`?

sebastinez replied:
28746cd

rudolfs commented on src/base/vesting/vesting.ts:
```diff
@@ -37,50 +46,88 @@ export async function withdrawVested(
+  let token: string | undefined = undefined;
+
+  try {
+    token = (await contract.token()) as string;
```
Shouldn't we handle the `undefined` case here instead of asserting that it's always a `string`?

<img width="243" alt="Screenshot 2022-12-23 at 09 59 12" src="https://user-images.githubusercontent.com/158411/209305648-8f3cb68c-7e53-4a60-ba7b-520768317a3e.png">

rudolfs commented on src/base/vesting/vesting.ts:
```diff
@@ -37,50 +46,88 @@ export async function withdrawVested(
+  let token: string | undefined = undefined;
+
+  try {
+    token = (await contract.token()) as string;
```
Could we move this out of the try block and just do:
```
token = await contract.token();
if(!token) {
return undefined;
}
```

Instead of the type assertion? I would find that easier to read.

sebastinez replied:
Hmm the reason I left it inside the `try` block is to catch any errors that could happen when fetching it.
I think I could leave it inside the `try` block but do the `if(!token)` guard to avoid the casting to `string`

sebastinez replied:
Something like
```js
  let token: string | undefined = undefined;

  try {
    token = await contract.token();
    if (!token) {
      return undefined;
    }
```

rudolfs commented on src/base/vesting/vesting.ts:
```diff
@@ -37,50 +46,88 @@ export async function withdrawVested(
   };
 }
+
+export const parseVestingPeriods = (...timestamps: string[]): string => {
```
```suggestion
export const parsedVestingPeriods = (...timestamps: string[]): string => {
```

brandonhaslegs commented:
This whole page is about vesting right? I don’t think you need to hide anything behind a button. What else is there to do on this page?

sebastinez commented:
> This whole page is about vesting right? I don’t think you need to hide anything behind a button. What else is there to do on this page?

The page is the profile page of a ethereum address, so there will be a listing of projects corresponding to a user or eventual org, and personal information like seed maybe balances for multi sig and eventually other stuff.

brandonhaslegs commented:
Or instead of just showing a button, can you show some value there?

So instead of:

**Vesting** [Show details]

you could indead show a value with a [View more] button next to it

**Vesting Beneficiary** 👤rudolfs.eth [View more]

And the “view more” button would then expand all the vesting info.

brandonhaslegs commented:
Things I think should change:
- The [back] button is weird. It should just timeout.
- The loading indicator looked better as the three-dot that’s everywhere else in the app
- The eth address after “Waiting for confirmation” should be truncated just like the address is truncated with the “Tokens have been withdrawn” success state. I made a separate issue for truncating all eth addresses here #555

brandonhaslegs commented:
I am also wondering if this withdrawl flow should be more “generic” like just exist in a modal, just in case you have to withdrawl from a different UI at some point. Then we could repurpose the UI.

sebastinez commented:
Thanks @brandonhaslegs for the feedback! Really appreciated!

> Or instead of just showing a button, can you show some value there?

That's a great idea, I'll think about some values we can show together with a [View more] toggle

> The [back] button is weird. It should just timeout.

Ok I'll try some other timeout durations, maybe we could also do something like when moving your mouse out of the textarea we timeout after some seconds, kinda like a debounce after user interaction? 

> The loading indicator looked better as the three-dot that’s everywhere else in the app

Got it

> The eth address after “Waiting for confirmation” should be truncated just like the address is truncated with the “Tokens have been withdrawn” success state.

The address shown on "Waiting for confirmation" is the transaction hash of the withdrawal, I thought it would be nice to give users this information in case something goes bad.
The transaction hash is trimmed to it's first 20 characters of 66 in total, I took the inspiration from etherscan but we can do it differently or don't show the tx hash at all 🤷 
<img width="475" alt="Bildschirm­foto 2022-12-23 um 22 40 46" src="https://user-images.githubusercontent.com/7912302/209407809-1af05dde-55b7-47de-bf26-21344a0db933.png">

> I am also wondering if this withdrawl flow should be more “generic” (for example just exist in a modal flow) just in case you have to withdrawl from a different UI at some point. Then we could repurpose the withdrawl UI. This would have to be designed of course, but we’ve designed many withdrawl flows at this point, I’m sure we could throw something together. Maybe this is something better to do in the future.

Really like the idea of a generic withdrawal component modal that can be shown at any point when there is some withdraw happening. Maybe that's something we can take into account in the modal refactor at #557

sebastinez commented:
Okay I addressed your feedback @brandonhaslegs and I'm open to review the etherscan icon https://github.com/radicle-dev/radicle-interface/pull/552#issuecomment-1363110256

I would like to get this PR closed before heartwood comes out.

rudolfs commented:
> I would like to get this PR closed before heartwood comes out.

You mean you want to merge it into master?

sebastinez commented:
> You mean you want to merge it into master?

Yeah exactly 👍

brandonhaslegs commented:
Here is a new etherscan icon

![etherscan](https://user-images.githubusercontent.com/4406983/209815106-0554185f-73b4-434e-ab37-73dbc9ee2541.svg)

brandonhaslegs commented:
Can you give me a URL where I can test this?

sebastinez commented:
> @sebastinez Can you give me a URL where I can test this?

Yeah sure, let me update the icon and I'll deploy it somewhere

brandonhaslegs commented:
Ah sorry I built it just fine from terminal, i just don’t know where to go to find this page

sebastinez commented:
> Ah sorry I built it just fine from terminal, i just don’t know where to go to find this page

Ok I just added a mocking function to test it more easily.
Pull the latest changes and try out this url http://localhost:3000/0x27BCA0692e13C122E6Fc105b3974B5df7246D464

sebastinez commented:
Or it can also be previewed here https://tubular-taiyaki-bf2c09.netlify.app/0x27BCA0692e13C122E6Fc105b3974B5df7246D464

rudolfs commented:
> I am also wondering if this withdrawl flow should be more “generic” (for example just exist in a modal flow) just in case you have to withdrawl from a different UI at some point. Then we could repurpose the withdrawl UI. This would have to be designed of course, but we’ve designed many withdrawl flows at this point, I’m sure we could throw something together. Maybe this is something better to do in the future.

@brandonhaslegs this is how it is implemented now actually and this PR changes it to not be in a modal.

See: https://app.radicle.xyz/vesting

Here are some sample Görli vesting contracts:
```
0x13b2Fc1f601Fb72b86BFAB59090f22bB6E73005A
0xAADcbc69f955523B0ff0A271229961E950538EbE
0x27BCA0692e13C122E6Fc105b3974B5df7246D464
0x9c882463B02221b0558112dec05F60D5B3D99b6a
```


Original:

<img width="1840" alt="Screenshot 2022-12-28 at 15 43 07" src="https://user-images.githubusercontent.com/158411/209829086-c713e192-1a09-4f23-9432-a46e64d41cda.png">

<img width="1840" alt="Screenshot 2022-12-28 at 15 43 43" src="https://user-images.githubusercontent.com/158411/209829121-96b5c260-6780-4b12-89b3-12fd26bef0bc.png">

brandonhaslegs commented:
Well why dont we keep this with a few changes?

sebastinez commented:
> Well why dont we keep this with a few changes?

The idea of implementing the whole vesting information into a Profile page came up during a discussion here: https://github.com/radicle-dev/radicle-interface/issues/526#issuecomment-1338144182

I think if we would like to keep everything as is, we could just redirect searches for vesting contracts to the modal view.
And improve the withdraw process, and error handling in the vesting modal. 🤔

sebastinez commented:
Ok I think this PR is still better than the current status quo.. but if anyone wants to take another stab at designing this or has already something, we can sync maybe over Discord and bring this PR over the finish line.

cloudhead commented:
Feedback here:
<img width="765" alt="image" src="https://user-images.githubusercontent.com/40774/211055568-c4b7ec4a-d8d8-4e97-bd27-9c74970d12b5.png">

cloudhead commented:
Furthermore:

*withdraw* button should not show if there's nothing to withdraw or if you're not the beneficiary.
If you're not connected, it should be disabled.

sebastinez commented:
Ok addressed the feedback by @cloudhead.

Here
<img width="928" alt="Bildschirm­foto 2023-01-09 um 15 37 33" src="https://user-images.githubusercontent.com/7912302/211340595-d8cae638-994b-4403-91ee-e2247d11f16a.png">

And here some videos of the withdrawal flow.
One receives an error during the transaction wait, and one completes sucessfully.

https://user-images.githubusercontent.com/7912302/211340818-f4d60593-7c62-41ec-9829-52be996a869c.mov

https://user-images.githubusercontent.com/7912302/211340828-392d434c-88f8-4541-825f-268085471280.mov

rudolfs commented:
The etherscan icon doesn't match the rest of the social-media link icons:
<img width="319" alt="Screenshot 2023-01-09 at 16 30 28" src="https://user-images.githubusercontent.com/158411/211345375-386b4568-c22b-48f5-99dc-318941f22cb5.png">

It also doesn't seem to be centered in the bounding box.
<img width="426" alt="Screenshot 2023-01-09 at 16 30 55" src="https://user-images.githubusercontent.com/158411/211345448-bbdcf42c-029f-4dea-9c3e-0276e5357374.png">

Here are all our other icons:
<img width="454" alt="Screenshot 2023-01-09 at 16 37 50" src="https://user-images.githubusercontent.com/158411/211346925-a27992a2-7127-4a94-afb4-c84cefc347a0.png">

rudolfs commented:
It's possible to submit the input via the <kbd>enter</kbd> key even when the button is disabled:
<img width="695" alt="Screenshot 2023-01-09 at 16 42 43" src="https://user-images.githubusercontent.com/158411/211348000-0a07dc7d-9fd8-4787-96c7-51320358134a.png">

We should not allow submit on keypress when the form isn't valid.

When there is no vesting contract, there are no errors in the console. We should adjust the error message to just "No vesting account found under this address."
<img width="659" alt="Screenshot 2023-01-09 at 16 45 08" src="https://user-images.githubusercontent.com/158411/211348599-cdb4bddd-0a76-450b-a256-a97f3bafeb55.png">

rudolfs commented:
Is there a reason why this address is clickable? We are already on the profile page of it, no?
<img width="1146" alt="Screenshot 2023-01-09 at 16 49 37" src="https://user-images.githubusercontent.com/158411/211349664-077b1e2c-31d6-4868-be8b-90c6469ed7d0.png">

cloudhead commented:
Looking better! A couple more things:

* usually we show errors using the error modal which is all red; we don't mix red error text in blue modals
* In the success case, we should show the withdrawal address inline, eg. "tokens withdrawn to [icon] xyz - abc"

rudolfs commented:
We should make the gap between the buttons a bit wider, like 1rem?
<img width="564" alt="Screenshot 2023-01-09 at 16 54 10" src="https://user-images.githubusercontent.com/158411/211350674-a933c2bb-e973-4f5e-9093-dd05e9c948ed.png">

rudolfs commented:
When funds have been withdrawn and I click on the address in the modal, I get an error:

<img width="1840" alt="Screenshot 2023-01-09 at 16 56 43" src="https://user-images.githubusercontent.com/158411/211351442-6a14a339-c045-4a1e-9ef4-774b45fc51e2.png">

<img width="1840" alt="Screenshot 2023-01-09 at 16 56 46" src="https://user-images.githubusercontent.com/158411/211351466-bc316bfd-ef43-4db9-beed-1f52775a96f2.png">

cloudhead commented:
> Is there a reason why this address is clickable? We are already on the profile page of it, no?

I think that's a separate issue no? ie. it affects all profiles.

cloudhead commented:
The etherscan icon is just bigger than the other icons for some reason.

brandonhaslegs commented:
Yes! I recently redesigned our icon set (refined them, made them more streamlined). Here are the Github and Twitter ones. And the 🔗  one too!
![github](https://user-images.githubusercontent.com/4406983/211378227-4c2331ca-7d37-46c5-ad4e-71addc432b81.svg)
![link](https://user-images.githubusercontent.com/4406983/211378229-2c88c707-682c-4711-95c5-2bdc6ad16b63.svg)
![twitter](https://user-images.githubusercontent.com/4406983/211378236-cd637ebc-76df-4824-91fa-192d55ce1639.svg)

sebastinez commented:
Ok did update the Icons.

<img width="668" alt="Bildschirm­foto 2023-01-09 um 22 07 17" src="https://user-images.githubusercontent.com/7912302/211408524-2e583fd8-0a16-4f4d-83ea-5fc32ff935a0.png">

Now etherscan surely seems more like the others.
@brandonhaslegs if you have other icons, for more streamlining please lmk and I'll do that in another PR
![image](https://user-images.githubusercontent.com/7912302/211408698-08c18d1c-c196-4b75-a8bc-19693d2c6fd9.png)
I think there are some others that could be improved

sebastinez commented:
> It's possible to submit the input via the enter key even when the button is disabled: <img alt="Screenshot 2023-01-09 at 16 42 43" width="695" src="https://user-images.githubusercontent.com/158411/211348000-0a07dc7d-9fd8-4787-96c7-51320358134a.png">

Good catch! 💪

sebastinez commented:
Ok addressed with the latest commit:

- the `ErrorModal` usage
- the withdrawal not working 😓 thanks for the catch @rudolfs!
- the error when clicking on the address where to tokens where withdrawn to.
- the ability to click enter and to trigger the `on:submit` of a `TextInput` component even if the input isn't valid.
- modified the error message, removing the dev console part.

A video as reference
https://user-images.githubusercontent.com/7912302/211507254-af37d79c-5c55-49b9-8a8c-9a1a88051e3f.mov

cloudhead commented:
Thanks, looking good, just one thing left: spacing/alignment here is off:
<img width="807" alt="image" src="https://user-images.githubusercontent.com/40774/211512767-92e53ed9-63c1-4ba2-816a-e9da64e07480.png">

I also wonder, what happens if you close the modal while it's confirming?

sebastinez commented:
> Thanks, looking good, just one thing left: spacing/alignment here is off: <img alt="image" width="807" src="https://user-images.githubusercontent.com/40774/211512767-92e53ed9-63c1-4ba2-816a-e9da64e07480.png">
> 
> I also wonder, what happens if you close the modal while it's confirming?

If you close the modal while confirming, after having signed the tx, the tx goes out either way and you loose eventually feedback if something goes wrong.
We could eventually just show the `Close` button once the funds have been withdrawn 🤔 but I wanted to give users the ability to close the modal also during the tx.

sebastinez commented:
> Thanks, looking good, just one thing left: spacing/alignment here is off

@cloudhead Ok I fixed this by removing the `Address` component here, and do it as we do in the registration flow, where we just show the formatted address.

<img width="530" alt="Bildschirm­foto 2023-01-10 um 10 34 25" src="https://user-images.githubusercontent.com/7912302/211514983-23654894-9ca1-442a-a2b8-f2dc28659076.png">

rudolfs commented:
The `withdrawn` amount doesn't update automatically after the tx is confirmed. It only updates after a page refresh.

rudolfs commented on src/lib/utils.ts:
```diff
@@ -290,10 +297,11 @@ export function isFulfilled<T>(
 
 // Get the explorer link of an address, eg. Etherscan.
 export function explorerLink(addr: string, wallet: Wallet): string {
```
We should rename the `addr` attribute to `addrOrTx`.

sebastinez replied:
1e4ebf0

rudolfs commented on src/lib/utils.ts:
```diff
@@ -310,6 +318,19 @@ export function parseRadicleId(id: string): string {
 
+// Checks that the input is an object with a `reason` property.
+// Is used to check if ether.js errors have a `reason` property.
+export function hasReasonProperty(
```
Were using this only in one place, could we inline it for now?

sebastinez replied:
1e4ebf0

sebastinez commented:
> The `withdrawn` amount doesn't update automatically after the tx is confirmed. It only updates after a page refresh.

1e4ebf0

rudolfs commented on src/lib/vesting.ts:
```diff
@@ -37,50 +47,113 @@ export async function withdrawVested(
+        contract.cliffPeriod(),
+      ]);
+    } catch (e) {
+      return undefined;
```
Can we log this error with `console.warn`?

sebastinez replied:
16ead15

rudolfs commented on src/lib/vesting.ts:
```diff
@@ -37,50 +47,113 @@ export async function withdrawVested(
+  { max: 1000 },
+);
+
+export const parseVestingPeriods = (...timestamps: string[]): string => {
```
Could we use a normal function definition instead of const where possible?

sebastinez replied:
16ead15

rudolfs commented on src/views/vesting/Withdraw.svelte:
```diff
@@ -0,0 +1,72 @@
+
+    <span slot="subtitle">
+      {#if $state.type === "withdrawingSign"}
+        <span class="txt-missing">Waiting for a signature...</span>
```
Could we use the ellipsis character … instead of `...` everywhere?

sebastinez replied:
5d2cbd3

rudolfs reviewed (approved):
👏 🚀
